rclcpp/include/rclcpp/strategies/msg_pool_memory_strategy.hpp

```diff
@@ -0,0 +1,80 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_RCLCPP_MSG_POOL_MEMORY_STRATEGY_HPP_
+#define RCLCPP_RCLCPP_MSG_POOL_MEMORY_STRATEGY_HPP_
+
+#include <rclcpp/msg_memory_strategy.hpp>
+#include <rclcpp/macros.hpp>
+
+namespace rclcpp
+{
+namespace msg_pool_memory_strategy
+{
+
+template<typename MessageT, size_t size>
+class MessagePoolMemoryStrategy
+  : public std::enable_if<rosidl_generator_traits::has_fixed_size<MessageT>::value,
+  msg_memory_strategy::MessageMemoryStrategy<MessageT>>::type
+{
+public:
+  RCLCPP_MAKE_SHARED_DEFINITIONS(MessagePoolMemoryStrategy);
+  MessagePoolMemoryStrategy()
+  : seq_(0)
+  {
+    for (size_t i = 0; i < size; ++i) {
+      msg_ptrs_[i] = std::make_shared<MessageT>(msg_pool_[i]);
+    }
+
+    msg_used_.fill(false);
+  }
+
+  std::shared_ptr<MessageT> borrow_message()
+  {
+    seq_ = (seq_ + 1) % size;
+    if (msg_used_[seq_]) {
+      throw std::runtime_error("Tried to access message that was still in use! Abort.");
+    }
+    msg_ptrs_[seq_]->~MessageT();
+    new (msg_ptrs_[seq_].get())MessageT;
+
+    msg_used_[seq_] = true;
+    return msg_ptrs_[seq_];
+  }
+
+  void return_message(std::shared_ptr<MessageT> & msg)
+  {
+    size_t i = 0;
+    for (; i < size; ++i) {
+      if (msg_ptrs_[i] == msg) {
+        break;
+      }
+    }
+    if (i >= size) {
+      throw std::runtime_error("Unrecognized message ptr in return_message.");
+    }
+
+    msg_used_[i] = false;
```

Comment:
To simplify the logic you could move the assignment into the condition in the loop and return instead of break. Then the exception is always thrown after the loop. And `i` can be declared within the for loop.